
	This patch brings over a few features from MirBSD:
	* -fhonour-copts
	  If this option is not given, it's warned (depending
	  on environment variables). This is to catch errors
	  of misbuilt packages which override CFLAGS themselves.
	* -Werror-maybe-reset
	  Has the effect of -Wno-error if GCC_NO_WERROR is
	  set and not '0', a no-operation otherwise. This is
	  to be able to use -Werror in "make" but prevent
	  GNU autoconf generated configure scripts from
	  freaking out.
	* Make -fno-strict-aliasing and -fno-delete-null-pointer-checks
	  the default for -O2/-Os, because they trigger gcc bugs
	  and can delete code with security implications.

	This patch was authored by Thorsten Glaser <tg@mirbsd.de>
	with copyright assignment to the FSF in effect.

Index: gcc-4.2.3/gcc/c-opts.c
===================================================================
--- gcc-4.2.3.orig/gcc/c-opts.c	2008-01-27 19:36:59.000000000 +0100
+++ gcc-4.2.3/gcc/c-opts.c	2008-05-21 13:46:01.550289703 +0200
@@ -106,6 +106,9 @@
 /* Number of deferred options scanned for -include.  */
 static size_t include_cursor;
 
+/* Check if a port honours COPTS.  */
+static int honour_copts = 0;
+
 static void set_Wimplicit (int);
 static void handle_OPT_d (const char *);
 static void set_std_cxx98 (int);
@@ -450,6 +453,14 @@
       mesg_implicit_function_declaration = 2;
       break;
 
+    case OPT_Werror_maybe_reset:
+      {
+	char *ev = getenv ("GCC_NO_WERROR");
+	if ((ev != NULL) && (*ev != '0'))
+	  cpp_opts->warnings_are_errors = 0;
+      }
+      break;
+
     case OPT_Wformat:
       set_Wformat (value);
       break;
@@ -692,6 +703,12 @@
       flag_exceptions = value;
       break;
 
+    case OPT_fhonour_copts:
+      if (c_language == clk_c) {
+	honour_copts++;
+      }
+      break;
+
     case OPT_fimplement_inlines:
       flag_implement_inlines = value;
       break;
@@ -1157,6 +1174,47 @@
   /* Has to wait until now so that cpplib has its hash table.  */
   init_pragma ();
 
+  if (c_language == clk_c) {
+    char *ev = getenv ("GCC_HONOUR_COPTS");
+    int evv;
+    if (ev == NULL)
+      evv = -1;
+    else if ((*ev == '0') || (*ev == '\0'))
+      evv = 0;
+    else if (*ev == '1')
+      evv = 1;
+    else if (*ev == '2')
+      evv = 2;
+    else if (*ev == 's')
+      evv = -1;
+    else {
+      warning (0, "unknown GCC_HONOUR_COPTS value, assuming 1");
+      evv = 1; /* maybe depend this on something like MIRBSD_NATIVE?  */
+    }
+    if (evv == 1) {
+      if (honour_copts == 0) {
+	error ("someone does not honour COPTS at all in lenient mode");
+	return false;
+      } else if (honour_copts != 1) {
+	warning (0, "someone does not honour COPTS correctly, passed %d times",
+	 honour_copts);
+      }
+    } else if (evv == 2) {
+      if (honour_copts == 0) {
+	error ("someone does not honour COPTS at all in strict mode");
+	return false;
+      } else if (honour_copts != 1) {
+	error ("someone does not honour COPTS correctly, passed %d times",
+	 honour_copts);
+	return false;
+      }
+    } else if (evv == 0) {
+      if (honour_copts != 1)
+	inform ("someone does not honour COPTS correctly, passed %d times",
+	 honour_copts);
+    }
+  }
+
   return true;
 }
 
Index: gcc-4.2.3/gcc/c.opt
===================================================================
--- gcc-4.2.3.orig/gcc/c.opt	2007-09-01 17:28:30.000000000 +0200
+++ gcc-4.2.3/gcc/c.opt	2008-05-21 13:46:01.550289703 +0200
@@ -188,6 +188,10 @@
 C ObjC RejectNegative
 Make implicit function declarations an error
 
+Werror-maybe-reset
+C ObjC C++ ObjC++
+; Documented in common.opt
+
 Wfloat-equal
 C ObjC C++ ObjC++ Var(warn_float_equal)
 Warn if testing floating point numbers for equality
@@ -543,6 +547,9 @@
 fhonor-std
 C++ ObjC++
 
+fhonour-copts
+C ObjC C++ ObjC++ RejectNegative
+
 fhosted
 C ObjC
 Assume normal C execution environment
Index: gcc-4.2.3/gcc/common.opt
===================================================================
--- gcc-4.2.3.orig/gcc/common.opt	2008-01-27 19:36:59.000000000 +0100
+++ gcc-4.2.3/gcc/common.opt	2008-05-21 13:46:01.550289703 +0200
@@ -80,6 +80,10 @@
 Common Joined
 Treat specified warning as error
 
+Werror-maybe-reset
+Common
+If environment variable GCC_NO_WERROR is set, act as -Wno-error
+
 Wextra
 Common
 Print extra (possibly unwanted) warnings
@@ -480,6 +484,9 @@
 Common Report Var(flag_guess_branch_prob)
 Enable guessing of branch probabilities
 
+fhonour-copts
+Common RejectNegative
+
 ; Nonzero means ignore `#ident' directives.  0 means handle them.
 ; Generate position-independent code for executables if possible
 ; On SVR4 targets, it also controls whether or not to emit a
Index: gcc-4.2.3/gcc/opts.c
===================================================================
--- gcc-4.2.3.orig/gcc/opts.c	2008-01-27 19:36:59.000000000 +0100
+++ gcc-4.2.3/gcc/opts.c	2008-05-21 13:46:01.562288718 +0200
@@ -483,9 +483,6 @@
       flag_schedule_insns_after_reload = 1;
 #endif
       flag_regmove = 1;
-      flag_strict_aliasing = 1;
-      flag_strict_overflow = 1;
-      flag_delete_null_pointer_checks = 1;
       flag_reorder_blocks = 1;
       flag_reorder_functions = 1;
       flag_tree_store_ccp = 1;
@@ -501,6 +498,10 @@
 
   if (optimize >= 3)
     {
+      flag_strict_aliasing = 1;
+      flag_strict_overflow = 1;
+      flag_delete_null_pointer_checks = 1;
+
       flag_inline_functions = 1;
       flag_unswitch_loops = 1;
       flag_gcse_after_reload = 1;
@@ -702,6 +703,17 @@
       }
       break;
 
+    case OPT_Werror_maybe_reset:
+      {
+	char *ev = getenv ("GCC_NO_WERROR");
+	if ((ev != NULL) && (*ev != '0'))
+	  warnings_are_errors = 0;
+      }
+      break;
+
+    case OPT_fhonour_copts:
+      break;
+
     case OPT_Wextra:
       set_Wextra (value);
       break;
Index: gcc-4.2.3/gcc/doc/cppopts.texi
===================================================================
--- gcc-4.2.3.orig/gcc/doc/cppopts.texi	2006-08-16 22:36:23.000000000 +0200
+++ gcc-4.2.3/gcc/doc/cppopts.texi	2008-05-21 13:46:01.570286943 +0200
@@ -166,6 +166,11 @@
 Make all warnings into hard errors.  Source code which triggers warnings
 will be rejected.
 
+@item -Werror-maybe-reset
+@opindex Werror-maybe-reset
+Act like @samp{-Wno-error} if the @env{GCC_NO_WERROR} environment
+variable is set to anything other than 0 or empty.
+
 @item -Wsystem-headers
 @opindex Wsystem-headers
 Issue warnings for code in system headers.  These are normally unhelpful
Index: gcc-4.2.3/gcc/doc/invoke.texi
===================================================================
--- gcc-4.2.3.orig/gcc/doc/invoke.texi	2008-05-21 13:45:54.274288067 +0200
+++ gcc-4.2.3/gcc/doc/invoke.texi	2008-05-21 13:46:01.582288192 +0200
@@ -226,7 +226,7 @@
 -Wc++-compat -Wcast-align  -Wcast-qual  -Wchar-subscripts  -Wcomment @gol
 -Wconversion  -Wno-deprecated-declarations @gol
 -Wdisabled-optimization  -Wno-div-by-zero  -Wno-endif-labels @gol
--Werror  -Werror=* -Werror-implicit-function-declaration @gol
+-Werror -Werror-maybe-reset -Werror=* -Werror-implicit-function-declaration @gol
 -Wfatal-errors  -Wfloat-equal  -Wformat  -Wformat=2 @gol
 -Wno-format-extra-args -Wformat-nonliteral @gol
 -Wformat-security  -Wformat-y2k @gol
@@ -3566,6 +3566,22 @@
 @option{-W}@var{foo}.  However, @option{-Wno-error=}@var{foo} does not
 imply anything.
 
+@item -Werror-maybe-reset
+@opindex Werror-maybe-reset
+Act like @samp{-Wno-error} if the @env{GCC_NO_WERROR} environment
+variable is set to anything other than 0 or empty.
+
+@item -fhonour-copts
+@opindex fhonour-copts
+If @env{GCC_HONOUR_COPTS} is set to 1, abort if this option is not
+given at least once, and warn if it is given more than once.
+If @env{GCC_HONOUR_COPTS} is set to 2, abort if this option is not
+given exactly once.
+If @env{GCC_HONOUR_COPTS} is set to 0 or unset, warn if this option
+is not given exactly once.
+The warning is quelled if @env{GCC_HONOUR_COPTS} is set to @samp{s}.
+This flag and environment variable only affect the C language.
+
 @item -Wstack-protector
 @opindex Wstack-protector
 This option is only active when @option{-fstack-protector} is active.  It
@@ -4893,7 +4909,7 @@
 second branch or a point immediately following it, depending on whether
 the condition is known to be true or false.
 
-Enabled at levels @option{-O2}, @option{-O3}, @option{-Os}.
+Enabled at levels @option{-O3}.
 
 @item -fcse-follow-jumps
 @opindex fcse-follow-jumps
@@ -5011,7 +5027,7 @@
 @option{-fno-delete-null-pointer-checks} to disable this optimization
 for programs which depend on that behavior.
 
-Enabled at levels @option{-O2}, @option{-O3}, @option{-Os}.
+Enabled at levels @option{-O3}.
 
 @item -fexpensive-optimizations
 @opindex fexpensive-optimizations
@@ -5460,7 +5476,7 @@
 allowed to alias.  For an example, see the C front-end function
 @code{c_get_alias_set}.
 
-Enabled at levels @option{-O2}, @option{-O3}, @option{-Os}.
+Enabled at levels @option{-O3}.
 
 @item -fstrict-overflow
 @opindex fstrict-overflow
Index: gcc-4.2.3/gcc/java/jvspec.c
===================================================================
--- gcc-4.2.3.orig/gcc/java/jvspec.c	2007-08-31 10:27:50.000000000 +0200
+++ gcc-4.2.3/gcc/java/jvspec.c	2008-05-21 13:46:01.582288192 +0200
@@ -631,6 +631,7 @@
      class name.  Append dummy `.c' that can be stripped by set_input so %b
      is correct.  */ 
   set_input (concat (main_class_name, "main.c", NULL));
+  putenv ("GCC_HONOUR_COPTS=s"); /* XXX hack!  */
   err = do_spec (jvgenmain_spec);
   if (err == 0)
     {
